lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp

```diff
@@ -2952,6 +2952,114 @@ static bool lowerRawSILOperations(SILFunction &Fn) {
   return Changed;
 }
 
+static SILBasicBlock *getOptionalDiamondSuccessor(SwitchEnumInst *SEI) {
+   auto numSuccs = SEI->getNumSuccessors();
+   if (numSuccs != 2)
+     return nullptr;
+   auto *SuccSome = SEI->getCase(0).second;
+   auto *SuccNone = SEI->getCase(1).second;
+   if (SuccSome->args_size() != 1)
+     std::swap(SuccSome, SuccNone);
+
+   if (SuccSome->args_size() != 1 || SuccNone->args_size() != 0)
+     return nullptr;
+
+   auto *Succ = SuccSome->getSingleSuccessorBlock();
+   if (!Succ)
+     return nullptr;
+
+   if (SuccNone == Succ)
+     return Succ;
+
+   SuccNone = SuccNone->getSingleSuccessorBlock();
+   if (SuccNone == Succ)
+     return Succ;
+
+   SuccNone = SuccNone->getSingleSuccessorBlock();
+   if (SuccNone == Succ)
+     return Succ;
+
+   return nullptr;
+}
+/// Ensure the lifetime of the closure accross an
+///
+///   optional<@escaping () -> ()> to
+///   optional<@noescape @convention(block) () -> ()>
+///
+///   conversion and its use.
+///
+///   The pattern this is looking for
+///                            switch_enum %closure
+///                           /           \
+///     convert_escape_to_noescape          nil
+///                             switch_enum
+///                           /           \
+///                convertToBlock          nil
+///                           \            /
+///                     (%convertOptionalBlock :)
+///   We will insert a copy_value of the original %closure before the two
+///   diamonds. And a destroy of %closure at the last destroy of
+///   %convertOptionalBlock.
+static bool fixupConvertEscapeToNoEscapeLifetime(SILFunction &Fn) {
```

Comment:
Ok, but *why* are you doing this transformation? Please explain in the comment why the SIL input to DI is invalid.